### ISSUE TITLE
How does the guideline ensure text is sufficiently readable for people with CVD (colourblindness)

### ISSUE THREAD

**Issue by sdw32** [w3c/silver#333]
In the current draft guideline, passing the 'visual contrast of text' guideline appears to be based solely on the score from the APCA tool. However, a quick investigation suggested it was possible to pass the APCA tool for text that became almost invisible with CVD

For example, if I enter #f00 foreground and #000 background for red text on a black background, APCA says I can pass at score 4 with 400-weight text at size 48px.  Yet if I screenshot this colour combination and feed it into the myndex CVD simulator, it is clear the red text becomes almost invisible, as per the screenshot below.
![Untitled-2](https://user-images.githubusercontent.com/79691399/109296496-49603980-7828-11eb-88d8-2e4feaacebb6.png)

I wondered if it would be possible to combine these 2 tools, so that the APCA tool runs the foreground and background colour combination through the CVD simulator, and reports the minimum of the Luminance contrasts from all the different types of colourblindness?

**Comment by Myndex** [comment]
@sdw32 Thank you for commenting

## Red at Night....
Pure red `#f00` is one of several known edge cases. It affects protans (~1% of population). In the guidelines it is discussed SEPARATELY from the APCA tool.

Nevertheless, It is NOT considered a "pass" in the APCA tool. With the G series constants, it is L<sup>c</sup> -37, below the L<sup>c</sup>45 cutoff for CONTENT (fluent readability) text. Not all of this information has migrated into the FPWD, and some aspects are still in progress.

The public CVD tool does not show at full resolution, so the perception is slightly skewed, as the version you are showing is at 1/3rd the actual resolution so the sample fonts are also 1/3rd the size they need to be. Nevertheless, **yes, pure red is difficult to read for a protan,** even in the full res version:

<img width="943" alt="sRGBredProtan" src="https://user-images.githubusercontent.com/42009457/109386565-545eac80-78b0-11eb-97e2-0738c93a92b8.png">


## Fixing Red for Protan: APCA Extension

There is already a color extension invented and tested to handle things like this, and the APCA algorithm is designed specifically to be extensible, so that features like this can be added. It is NOT added yet, due to IP reasons, and I can not discuss it any further publicly at this time.

### In the interim

The Visual Contrast _GUIDELINE_ discusses how pure red is prohibited as the brightest of two colors. The APCA tool reports it at under the cutoff of 45 (G series constants) intended for content text. And to note, this is an extreme edge case, applying only to protanopia.

### WCAG 2.x vs APCA
And even without the color module, the APCA is a massive improvement here from WCAG 2.x. Take a look at how WCAG 2.x rates that same pure red:

<img width="690" alt="WCAG 2 x red on black" src="https://user-images.githubusercontent.com/42009457/109385935-1fe8f180-78ac-11eb-945d-ba8d9d1dfdc4.png">

WCAG 2.x gives this a 5.25:1, which passes ANY size text at the AA level. As a side note, the WCAG 2.x was intended to "help" protans, though it actually makes some color combinations worse for them.

This and other WCAG 2.x comparisons are discussed in these two brief articles:

[**_Orange you Wondering About Contrast?_**
](https://gist.github.com/Myndex/1dadb6dcac596f1cd7a5686a076f697f)

[**_The Lighter Side of Dark Backgrounds_**](https://gist.github.com/Myndex/c30dba273aa5eca426ad9f5200917c9d)


Thank you,

Andy


_Andrew Somers_
_W3 AGWG Invited Expert_
_Myndex Color Science Researcher_
https://www.myndex.com/APCA/simple


Addendum: This is that same pure red slide, normal vision:

<img width="1120" alt="Screen Shot 2021-02-27 at 2 32 26 AM" src="https://user-images.githubusercontent.com/42009457/109386264-6b040400-78ae-11eb-9f5a-379a89fc0834.png">

**Comment by ChrisLoiselle** [comment]
@jspellman this is ready for survey.

**Comment by ChrisLoiselle** [comment]
DRAFT RESPONSE:  We believe the questions were answered per https://github.com/w3c/silver/issues/333#issuecomment-787063163.

I hope this answers your question, if it does not, please feel free to follow-up.

Thank you,
Chris

**Comment by sdw32** [comment]
Yes the latest algorithms handle red on black better than WCAG 2.X, but yet there are still 'edge' cases for colour combinations that the current APCA algorithm will pass for certain use cases, yet become problematic with Protanopia. 

@Myndex thanks for your comments. I had a look at some colours with a heavy red component that pass a score of 60, and yes in this case the effect of the protanopia is not as severe. 

However, the problem can still occur  for other use cases that can pass with a lower score. So, even in the latest tool that I tried today, RGB(255, 0, 0) gives a score of  38, and this would pass for text for spot reading (like a label in a diagram)

Running this colour through your simulator gives an equivalent of RGB (96, 96, 15) with simulated protanopia, and running this colour through the APCA tool gives a score of 19, which would not be suitable for this use case.

So, the issue can still occur where certain combinations pass the current APCA tool yet are considerably inappropriate with protanopia.

From conversations with my current clients, they are looking for a single catch all test that covers both CVD and contrast, rather than 2 separate tests. For a particular colour combination, I believe this could be achieved by: 
*running the RGB set 'as is' through the APCA algorithm
*modifying the RGB set to simulate protanopia, and then run it through the APCA algorithm again
*report the worst score that comes out of this
this approach would be readily extensible to simulate other kinds of CVD as well. Achieving this would require a public domain RGB protanopia transformation algorithm. I don't know if such a thing exists, but it would be ideal if it did!

**Comment by Myndex** [comment]
### UPDATE

> [!NOTE]
> A version set to release this year which includes the protan compensator module. The protan compensator (recited in the published patent app) directly addresses the contrast deficiencies that protan-types experience when colors with a large rad component are the lightest of a color pair.
> 
> This will become more important for wider gamut color spaces, especially UHD (Rec2100) where the red primary is outside the visible range of protanopia.

**Comment by sdw32** [comment]
@Myndex Great, Will this be embedded within a freely available tool?  I would be glad to help publicise this once it is available

**Comment by Myndex** [comment]
> [@Myndex](https://github.com/Myndex) Great, Will this be embedded within a freely available tool? I would be glad to help publicise this once it is available

Yes, it will be releasing with the next draft of the [APCA Readability Criterion](https://readtech.org/ARC/#visual-contrast-text)

**Comment by fstrr** [comment]
Proposed response:

This content has been updated with the release of the latest draft of WCAG 3, which can be found here:  [W3C Accessibility Guidelines \(WCAG\) 3.0](https://www.w3.org/TR/wcag-3.0/).  We believe that these issues are no longer relevant based on the changes made to the latest working draft.  Should you feel that these issues remain in the current draft, please log a new issue here:  [New Issue · w3c/wcag3](https://github.com/w3c/wcag3/issues/new).
